### PR TITLE
Organizations api - Closes #10

### LINKED ISSUE #10: Organizations API

### PR DESCRIPTION
**Pull request by rodri042**
Fixes https://github.com/mumuki/mumuki-office/issues/10
@aguspina can you review it please?

### CODE REVIEW

File: app/controllers/api/organizations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    def protect_for_janitor!(organization)
+      protect! :janitor, organization.slug
+    end
+
+    def protect_for_owner!(organization)
+      protect! :owner, organization.slug
+    end
+
+    def protect_for_god!
+      protect! :owner, 'academy/*'
```
Comment (aguspina) [inline app/controllers/api/organizations_controller.rb]: I do not think this is a good idea, i would protect create with `protect_for_owner` method

File: app/controllers/api/organizations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    end
+
+    def protect!(role, slug)
+      @api_client.protect! role, slug
```
Comment (aguspina) [inline app/controllers/api/organizations_controller.rb]: The protect logic is already in `CoursesController` may be, you could add a concern

File: README.md [hunk 3]
```diff
@@ -317,9 +317,36 @@ DELETE /courses/:id
...
+
+## Mandatory fields
+```json
+{
+  "name": "academy",
+  "contact_email": "issues@mumuki.io",
+  "books": [
+    "MumukiProject/mumuki-libro-metaprogramacion"
+  ],
+  "login_methods": [
```
Comment (flbulgarelli) [inline README.md]: Actually login methods are optional, and apply only for auth providers with social capabilities - auth0 only currently.

File: README.md [hunk 4]
```diff
@@ -329,32 +356,44 @@ get /organizations
...
 ```json
-{
-  "organizations": [
-    {"name": "academy", "logo_url":"http://...", "private": false},
-    {"name": "alcal", "private": true}
-  ]
-}
+[
+  { "name": "academy", "contact_email": "a@a.com", "locale": "es-AR", "login_methods": ["facebook"], "books": ["libro"], "private": false, "logo_url":"http://..." },
+  { "name": "alcal", "contact_email": "b@b.com", "locale": "en-US", "login_methods": ["facebook", "github"], "books": ["book"], "private": true }
```
Comment (flbulgarelli) [inline README.md]: :warning: only the following organizations should be returned: 

* public ones
* private organization you have access to. So it is not simply `**Minimal permission**: None.`

File: spec/factories/api_clients.rb [hunk 5]
```diff
@@ -1,6 +1,11 @@
...
-  factory :api_client do
+  factory :api_client do |t|
+    transient do
+      role :janitor
+      grant "test/*"
+    end
+
     description "foo"
-    user { create :user, first_name: 'foo', last_name: 'bar', email: 'foo+1@bar.com', permissions: {janitor: 'test/*'} }
+    user { create :user, first_name: 'foo', last_name: 'bar', email: 'foo+1@bar.com', permissions: Hash[role, grant] }
```
Comment (flbulgarelli) [inline spec/factories/api_clients.rb]: Don't use this syntax. Use the standard `{ role => grant }`  form

File: db/schema.rb [hunk 6]
```diff
@@ -40,14 +40,18 @@
   create_table "organizations", force: :cascade do |t|
     t.string   "name"
     t.text     "description"
-    t.string   "book_slug"
     t.string   "logo_url"
-    t.string   "login_methods"
-    t.boolean  "private"
+    t.boolean  "private",          default: false
```
Comment (flbulgarelli) [inline db/schema.rb]: New organizations should be private by default. Perhaps we could change this field semantics to `public`, so default is still false.

File: db/migrate/20170113202636_make_login_methods_an_array_on_organization.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class MakeLoginMethodsAnArrayOnOrganization < ActiveRecord::Migration[5.0]
+  def change
+    change_table :organizations do |t|
+      t.remove :login_methods
+      t.string :login_methods, array: true, default: []
+    end
```
Comment (flbulgarelli) [inline db/migrate/20170113202636_make_login_methods_an_array_on_organization.rb]: I think that you should use `text` instead of `string`, since `string` is limited to 255 chars.

File: db/migrate/20170113064743_add_books_locale_and_tos_to_organization.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+class AddBooksLocaleAndTosToOrganization < ActiveRecord::Migration[5.0]
+  def change
+    change_table :organizations do |t|
+      t.remove :book_slug
+      t.string :books, array: true, default: []
+
+      t.string :locale
+      t.string :terms_of_service
```
Comment (flbulgarelli) [inline db/migrate/20170113064743_add_books_locale_and_tos_to_organization.rb]: Terms of service should be `text`, not `string`

File: db/migrate/20170113064743_add_books_locale_and_tos_to_organization.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+class AddBooksLocaleAndTosToOrganization < ActiveRecord::Migration[5.0]
+  def change
+    change_table :organizations do |t|
+      t.remove :book_slug
+      t.string :books, array: true, default: []
```
Comment (flbulgarelli) [inline db/migrate/20170113064743_add_books_locale_and_tos_to_organization.rb]: Should be `text`

File: config/boot.rb [hunk 10]
```diff
@@ -1,3 +1,14 @@
...
 
 require 'bundler/setup' # Set up gems listed in the Gemfile.
+
+require 'rails/commands/server' 
+
+module Rails
+  class Server
+    alias :default_options_alias :default_options
+    def default_options
+      default_options_alias.merge!(:Host => '0.0.0.0')
```
Comment (flbulgarelli) [inline config/boot.rb]: Why is this necessary?

File: app/models/organization.rb [hunk 11]
```diff
@@ -1,2 +1,20 @@
 class Organization < ApplicationRecord
+  validates :name, :uniqueness => true, :format => { :with => /\A[-a-z0-9_]*\z/ }
```
Comment (flbulgarelli) [inline app/models/organization.rb]: When key is a symbol, use `key: value` instead of `:key => value`

File: app/models/organization.rb [hunk 12]
```diff
@@ -1,2 +1,20 @@
...
+  validates_presence_of :name, :contact_email, :locale
+  validates :books, :at_least_one => true
+  validates :login_methods, :at_least_one => true
+  validates :locale, :inclusion => { :in => %w(es-AR en-US) }
+  before_save :default_values
+
+  def slug
+    Mumukit::Auth::Slug.join self.name
```
Comment (flbulgarelli) [inline app/models/organization.rb]: Don't use `self` unless necessary

File: app/controllers/concerns/with_protection.rb [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  end
+
+  def protect!(role, slug)
+    @api_client.protect! role, slug
+  end
+
+  private
+
+  def set_api_client
+    @api_client = ApiClient.find_by!(token: authorization_header.split(' ').last)
```
Comment (flbulgarelli) [inline app/controllers/concerns/with_protection.rb]: Use `Mumukit::Auth::Token.extract_from_header`

File: app/controllers/concerns/with_custom_unhandled_errors.rb [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  included do
+    rescue_from Exception do |e| unhandled_error!(e) end
+  end
+
+  private
+
+  def unhandled_error!(e)
+    summary = {
+        errors: {
+            :exception => "#{e.class.name} : #{e.message}"
```
Comment (flbulgarelli) [inline app/controllers/concerns/with_custom_unhandled_errors.rb]: Use `exception: ...`

File: app/controllers/concerns/with_custom_unhandled_errors.rb [hunk 15]
```diff
@@ -0,0 +1,20 @@
+module WithCustomUnhandledErrors
+  extend ActiveSupport::Concern
+
+  included do
+    rescue_from Exception do |e| unhandled_error!(e) end
```
Comment (flbulgarelli) [inline app/controllers/concerns/with_custom_unhandled_errors.rb]: Use `{}`  instead of `do..end` for one-liners

**Comment by flbulgarelli** [comment]
Nice work @rodri042. Please review my comments. 

We need to create a consumer task in laboratory and classroom-api that listen to organization changes and update the organization accordingly. Could you do that too, please?

**Comment by rodri042** [comment]
@flbulgarelli 

Yep. That change should include creating the new organizations in their local databases?

File: README.md (hunk 3 above)
Comment (rodri042) [inline README.md]: So the default should be `'[]'`? Isnt `user_pass` a login method?

File: config/boot.rb (hunk 10 above)
Comment (rodri042) [inline config/boot.rb]: By default the server listen on 127.0.0.1 instead of 0.0.0.0 and then it's unreachable by the host machine running the vagrant vm. Another solution could be running always with `rails s -b 0.0.0.0

File: README.md (hunk 4 above)
Comment (rodri042) [inline README.md]: Question:
Is there any way to know if a permission has *any role* for a slug? I wouldn't like to ask for permissions with `:student` **AND** `:writer` [[link](https://yuml.me/diagram/plain/class/%5BOwner%5D%5E-%5BJanitor%5D,%20%5BJanitor%5D%5E-%5BHeadmaster%5D,%20%5BHeadmaster%5D%5E-%5BTeacher%5D,%20%5BTeacher%5D%5E-%5BStudent%5D,%20,%20%5BOwner%5D%5E-%5BEditor%5D,%20%5BEditor%5D%5E-%5BWriter%5D)]

**Comment by flbulgarelli** [comment]
> Yep. That change should include creating the new organizations in their local databases?

Yes. Laboratory actually handles organizations already, so it has an organization model. I think that classroom has - or at least, should have - an organization model. Bibliotheca and Office do not need them actually.

**Comment by rodri042** [comment]
@flbulgarelli please re-review

File: config/boot.rb (hunk 10 above)
Comment (flbulgarelli) [inline config/boot.rb]: Yes, I'd prefer configuring it from command line instead of tweaking the code in non-standard ways.

File: app/controllers/api/organizations_controller.rb [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+    def index
+      organizations = Organization.all.select do |it|
+        it.public || has_permission?(:janitor, it.slug)
+      end
+      render json: organizations
+    end
+
+    def show
+      organization = Organization.find_by_name id_param
```
Comment (flbulgarelli) [inline app/controllers/api/organizations_controller.rb]: :warning: `find_by_something` is depracated. Use `find_by something: value` instead.

File: app/controllers/api/organizations_controller.rb [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+    def create
+      protect_for_owner! Organization.new organization_params
+      organization = Organization.create! organization_params
+      organization.notify!
+      render json: organization
+    end
+
+    def update
+      organization = Organization.find_by_name id_param
```
Comment (flbulgarelli) [inline app/controllers/api/organizations_controller.rb]: Why the name of an organization is in an `id` field? It should be a `name` field

**Comment by flbulgarelli** [comment]
Ok, it looks good to me. Merging. Please have a look at my comments and changes, and try to fix them on next themes PR.

Please handle JS customization on another PR, too.

File: app/controllers/api/organizations_controller.rb (hunk 17 above)
Comment (rodri042) [inline app/controllers/api/organizations_controller.rb]: If anyone wants to change the `name` of an organization in `PUT /organizations/:name`, it would not work. Anyway, I don't know if we want to handle that situation

Comment (flbulgarelli) [inline app/controllers/api/organizations_controller.rb]: Ok, you have a point, but we don't want to handle that situation. Organization renaming is not supported, since it means changing lots of things.
